nan.h

```diff
@@ -1527,26 +1755,28 @@ class NanObjectWrap : public node::ObjectWrap {
 template <typename T, typename P>
 class NanWeakCallbackData {
  public:  // constructors
+  inline ~NanWeakCallbackData();
+
+ private:  // constructors
+  NAN_DISALLOW_ASSIGN_COPY_MOVE(NanWeakCallbackData)
+
+ public:
   typedef void (*Callback)(
       NanWeakCallbackData & data  // NOLINT(runtime/references)
       );
-  NanWeakCallbackData(v8::Handle<T> handle, P* param, Callback cb)
-    : parameter(param), callback(cb) {
-    NanAssignPersistent(persistent, handle);
-    Revive();
+  NanWeakCallbackData(v8::Persistent<T> *handle, P* param, Callback cb)
+    : persistent(), parameter(param), callback(cb) {
+    std::memcpy(&persistent, handle, sizeof(persistent));
```

Comment:
The Persistent<T>, the handle, is on the stack. It points to a T*, which is persistently stored. The handle is freed when it goes out of scope, but the T* remains and needs disposing later, in the weak callback. I can't dispose of the T* without a Persistent<T>, but I cannot construct a Persistent<T> from a T*, so the only way is by doing a bit copy of the original Persistent<T> to a heap-allocated struct.

On May 18, 2015 9:04:01 PM EEST, Trevor Norris <notifications@github.com> wrote:
>>    typedef void (*Callback)(
>>        NanWeakCallbackData & data  // NOLINT(runtime/references)
>>        );
>> -  NanWeakCallbackData(v8::Handle<T> handle, P* param, Callback cb)
>> -    : parameter(param), callback(cb) {
>> -    NanAssignPersistent(persistent, handle);
>> -    Revive();
>> +  NanWeakCallbackData(v8::Persistent<T> *handle, P* param, Callback
>cb)
>> +    : persistent(), parameter(param), callback(cb) {
>> +    std::memcpy(&persistent, handle, sizeof(persistent));
>
>To verify you need to store the `v8::Persistent<T> *handle` that's
>passed to `NanWeakCallbackData()`? IIRC the pointer to a
>`v8::Persistent` will never change in its lifetime. So couldn't you do
>something like:
>```cpp
>NanWeakCallbackData(v8::Persistent<T>* handle, P* pram, Callback cb)
>    : persistent(&handle),
>      parameter(param),
>      callback(cb) {
>}
>```
>So then this would change to use a pointer:
>```cpp
> private:  // data members
>  v8::Persistent<T>* persistent;
>```
>You can then continue to do operations like the following:
>```cpp
>  bool IsNearDeath() const { return (*persistent).IsNearDeath(); }
>```
>Though I could also be completely missing something.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/pull/348/files#r30532118